include/dali/c_api.h

```diff
@@ -16,221 +16,268 @@
+  DALI_FLOAT    =  9,
+  DALI_FLOAT64  =  10,
+  DALI_BOOL     =  11,
+} dali_data_type_t;
+
+/**
+ * @brief Create DALI pipeline. Setting batch_size,
+ * num_threads or device_id here overrides
+ * values stored in the serialized pipeline.
+ * When separated_execution is false, prefetch_queue_depth is considered,
+ * gpu_prefetch_queue_depth and cpu_prefetch_queue_depth are ignored.
+ * When separated_execution is true, cpu_prefetch_queue_depth and
+ * gpu_prefetch_queue_depth are considered and prefetch_queue_depth is ignored.
+ */
+DLL_PUBLIC void daliCreatePipeline(daliPipelineHandle *pipe_handle,
+                                   const char *serialized_pipeline,
+                                   int length,
+                                   int batch_size,
+                                   int num_threads,
+                                   int device_id,
+                                   bool separated_execution,
+                                   int prefetch_queue_depth,
+                                   int cpu_prefetch_queue_depth,
+                                   int gpu_prefetch_queue_depth);
+
+/// @{
+/**
+ * @brief Feed the data to ExternalSource as contiguous memory.
+ *
+ * When calling this function, you need to provide a CUDA stream, which will be used when
+ * copying data onto GPU. This function is asynchronous, so it's your responsibility to
+ * synchronize on a provided CUDA stream.
+ *
+ * Keep in mind, that for the special case, where the data exists on the CPU and the
+ * ExternalSource's Backend in also a CPU, stream is not needed - feel free to pass
+ * the default stream.
+ *
+ * A convenience, synchronous, overload function is provided,
+ * which handles the stream synchronization.
+ *
+ * @param pipe_handle Pointer to pipeline handle
+ * @param name Pointer to a null-terminated byte string with the name of the External Source
+ *             to be fed
+ * @param device Device of the supplied memory.
+ * @param data_ptr Pointer to contiguous buffer containing all samples
+ * @param data_type Type of the provided data
+ * @param shapes Pointer to an array containing shapes of all samples concatenated one after
+ *              another. Should contain batch_size * sample_dim elements.
+ * @param sample_dim The dimensionality of a single sample.
+ * @param layout_str Optional layout provided as a pointer to null-terminated byte string.
+ *                   Can be set to NULL.
+ * @param stream CUDA stream to use when copying the data onto GPU. Remember to synchronize on the
+ *               provided stream.
+ */
+DLL_PUBLIC void
+daliSetExternalInputAsync(daliPipelineHandle *pipe_handle, const char *name,
+                          device_type_t device, const void *data_ptr,
+                          dali_data_type_t data_type, const int64_t *shapes,
+                          int sample_dim, const char *layout_str,
+                          cudaStream_t stream);
+
+DLL_PUBLIC void
```

Comment:
I would put some remark that if the GPU memory is provided it need to be available on the GPU already as this variant uses internal, non default stream to copy it to DALI.